### ISSUE TITLE
Slow queries when reading in order

### ISSUE THREAD

**Issue by holycheater** [ClickHouse/ClickHouse#40583]
**Describe the situation**
We have a certain data-set with user logs, trying to get all data by certain tags spanning several hours.
Queries with ordering by part of ordering keys work slower than queries which don't use ordering by order-key

**How to reproduce**
Tried on ClickHouse 22.3 initially, same behavior on `22.7.4.16` too

CREATE statement:
```
CREATE TABLE logs_time_dt
(
    `time` DateTime64(9) Codec(Delta, ZSTD(7)),
    `project` LowCardinality(String) CODEC(ZSTD(7)),
    `service` LowCardinality(String) CODEC(ZSTD(7)),
    `message` String CODEC(ZSTD(7)),
    `tags_hash` Array(UInt64) CODEC(ZSTD(7)),
    INDEX idx_message message TYPE ngrambf_v1(3, 512, 2, 0) GRANULARITY 3,
    INDEX idx_tags_hash tags_hash TYPE bloom_filter(0.01) GRANULARITY 1
)
ENGINE = MergeTree
PARTITION BY toStartOfHour(time)
ORDER BY (project, service, time)
SETTINGS index_granularity = 1024
```

example dataset, several hours worth of data, the result set by particular tag is 20k rows:
```
insert into logs_time_dt
(time, project, service, message, tags_hash)
select
fromUnixTimestamp64Nano(toInt64(toUnixTimestamp64Nano(toDateTime64('2022-08-01',9))+number/(2777)*1e9)),
'test' as project,
'test' as service,
'foo',
[ number % 3000 ]
from system.numbers
limit 60*1e6
```

slow query reading by key in reverse order:
```
SELECT *
FROM logs_time_dt
WHERE (project = 'test') AND (service = 'test') AND has(tags_hash, 42)
ORDER BY time DESC
FORMAT `Null`

Query id: f3277eec-3a99-4442-849d-1209ad402548

Ok.

0 rows in set. Elapsed: 8.770 sec. Processed 29.27 million rows, 1.11 GB (3.34 million rows/s., 126.91 MB/s.)
```

reading in order seems to be faster in this particular example, but still slow:
```
SELECT *
FROM logs_time_dt
WHERE (project = 'test') AND (service = 'test') AND has(tags_hash, 42)
ORDER BY time ASC
FORMAT `Null`

Query id: a3ce44be-5ea2-4ffc-9c72-7f4fc656a5e1

Ok.

0 rows in set. Elapsed: 3.788 sec. Processed 29.27 million rows, 1.11 GB (7.73 million rows/s., 293.59 MB/s.)
```

sorting by negative timestamp works a lot faster with this case:
```
SELECT *
FROM logs_time_dt
WHERE (project = 'test') AND (service = 'test') AND has(tags_hash, 42)
ORDER BY -toUnixTimestamp64Nano(time) ASC
FORMAT `Null`

Query id: e5415b8d-556c-4ffa-910a-81902b252f55

Ok.

0 rows in set. Elapsed: 0.442 sec. Processed 29.27 million rows, 1.11 GB (66.28 million rows/s., 2.52 GB/s.)
```

**Expected performance**
I expect the "correct" query with ordering by primary key suffix wouldn't be slower than ordering by a computed expression.

**Comment by den-crane** [comment]
`ORDER BY project DESC,   service DESC,    time DESC`


```sql
ORDER BY time DESC
0 rows in set. Elapsed: 7.638 sec. Processed 29.27 million rows, 1.11 GB (3.83 million rows/s., 145.72 MB/s.)

SELECT *
FROM logs_time_dt
WHERE (project = 'test') AND (service = 'test') AND has(tags_hash, 42)
ORDER BY
    project DESC,
    service DESC,
    time DESC
FORMAT `Null`

0 rows in set. Elapsed: 0.965 sec. Processed 29.27 million rows, 1.11 GB (30.33 million rows/s., 1.15 GB/s.)
```

**Comment by den-crane** [comment]
you need to remove LowCardinality

```sql
CREATE TABLE logs_time_dt
(
    `time` DateTime64(9) CODEC(Delta, ZSTD(7)),
    `project` String CODEC(ZSTD(7)),
    `service` String CODEC(ZSTD(7)),
    `message` String CODEC(ZSTD(7)),
    `tags_hash` Array(UInt64) CODEC(ZSTD(7)),
    INDEX idx_message message TYPE ngrambf_v1(3, 512, 2, 0) GRANULARITY 3,
    INDEX idx_tags_hash tags_hash TYPE bloom_filter(0.01) GRANULARITY 1
)
ENGINE = MergeTree
PARTITION BY toStartOfHour(time)
ORDER BY (project, service, time)
SETTINGS index_granularity = 1024;


SELECT *
              FROM logs_time_dt
              WHERE (project = 'test') AND (service = 'test') AND has(tags_hash, 42)
              ORDER BY time DESC 
limit 10
              FORMAT `Null`
              ;
0 rows in set. Elapsed: 0.076 sec. Processed 112.02 thousand rows, 6.95 MB (1.48 million rows/s., 91.81 MB/s.)
```

**Comment by orloffv** [comment]
Sandbox with all queries and one simple query
```
SET optimize_read_in_order = 0;
```
https://fiddle.clickhouse.com/ba6e3a90-c41c-4680-ab77-06614f49cd6b
```
SET optimize_read_in_order = 1;
```
https://fiddle.clickhouse.com/dc407df3-099e-4515-beb5-832f619c75e7
<img width="1500" alt="Screen Shot 2022-08-24 at 21 49 05" src="https://user-images.githubusercontent.com/504986/186500667-75acc4cd-261a-4173-a5d4-cacf3c2c72dd.png">
<img width="1501" alt="Screen Shot 2022-08-24 at 21 48 14" src="https://user-images.githubusercontent.com/504986/186500701-6bc1392e-0215-4b14-90d8-b898b6984ed0.png">

**Comment by canhld94** [comment]
If you want to keep order by primary key, temporary set `optimize_read_in_order` to `0`. There's certainly a flaw in read in order optimization that make it worse in this case.

**Comment by devcrafter** [comment]
For a reason, reading in order is used for the slow queries, which is not beneficial for the queries. The workaround is setting `optimize_read_in_order=0`.

<pre>
SELECT *
FROM logs_time_dt
WHERE (project = 'test') AND (service = 'test') AND has(tags_hash, 42)
ORDER BY time ASC
<b>SETTINGS optimize_read_in_order = 0</b>
FORMAT `Null`

0 rows in set. Elapsed: <b>0.168 sec</b>. Processed 29.27 million rows, 1.11 GB (174.12 million rows/s., 6.62 GB/s.)
</pre>
`EXPLAIN PIPELINE` can be used to see if `reading in order` is used:
<pre>
EXPLAIN PIPELINE
SELECT *
FROM logs_time_dt
WHERE (project = 'test') AND (service = 'test') AND has(tags_hash, 42)
ORDER BY time ASC
<b>SETTINGS optimize_read_in_order = 0</b>

Query id: 78f9ca2a-8a87-42df-92ec-c144ec2cb330

┌─explain──────────────────────────────────┐
│ (Expression)                             │
│ ExpressionTransform                      │
│   (Sorting)                              │
│   MergingSortedTransform 16 → 1          │
│     MergeSortingTransform × 16           │
│       LimitsCheckingTransform × 16       │
│         PartialSortingTransform × 16     │
│           (Expression)                   │
│           ExpressionTransform × 16       │
│             (Filter)                     │
│             FilterTransform × 16         │
│               (ReadFromMergeTree)        │
│               <b>MergeTreeThread × 16 0 → 1 │</b>
└──────────────────────────────────────────┘
</pre>
<pre>
EXPLAIN PIPELINE
SELECT *
FROM logs_time_dt
WHERE (project = 'test') AND (service = 'test') AND has(tags_hash, 42)
ORDER BY time ASC
<b>SETTINGS optimize_read_in_order = 1</b>

Query id: bd609a03-c96a-4fed-8e10-03be4f921106

┌─explain─────────────────────────────┐
│ (Expression)                        │
│ ExpressionTransform                 │
│   (Sorting)                         │
│   MergingSortedTransform 46 → 1     │
│     (Expression)                    │
│     ExpressionTransform × 46        │
│       (Filter)                      │
│       FilterTransform × 46          │
│         (ReadFromMergeTree)         │
│         <b>MergeTreeInOrder × 46 0 → 1 │</b>
└─────────────────────────────────────┘
</pre>

**Comment by KochetovNicolai** [comment]
The best solution to me is not to disable `optimize_read_in_order`, but try to improve performance for such type of queries.
FYI: `query_plan_read_in_order` is a setting to change read-in-order analysis (to check if we want to apply optimization); execution was not changed.

In this case, likely, a query with `optimize_read_in_order = 0` works faster because of multi-threaded execution. I can expect that due to good filter condition the result set is small. Sorting of full result set can fast in this case, and reading in parallel is more efficient vs. almost single-threaded reading with `optimize_read_in_order = 1`. (It is my guess and good to check).

I think that we can make reading more parallelized when `read-in-order` enabled. We can try different approaches.

**First idea**
Let's try to add buffering processor into query pipeline, after `FilterTransform` and before `MergingSortedTransform`. It will
* if LIMIT is not specified for `MergingSortedTransform`, allow reading more data and buffer it (e.g. up to `max_block_size * 2` rows, or any other constant)
* if LIMIT is specified, the same, but formula will be like `min(LIMIT, max_block_size * 2)`

**Second idea**
Try to build more parallel pipeline. Maybe split ` MergeTreeInOrder × 46` input streams between different threads, and do `MergingSortedTransform` separately in every thread (also, final `MergingSortedTransform` will be needed). So that pipeline will be like:
```
MergingSortedTransform N → 1
  MergingSortedTransform M_1 → 1
    MergeTreeInOrder × M_1
...
  MergingSortedTransform M_N → 1
    MergeTreeInOrder × M_N
```
where `M_1 + ... + M_N = 46`.

Both ideas is not good for the case when there is a single sorted part.

**Comment by canhld94** [comment]
> In this case, likely, a query with `optimize_read_in_order = 0` works faster because of multi-threaded execution. I can expect that due to good filter condition the result set is small. Sorting of full result set can fast in this case, and reading in parallel is more efficient vs. almost single-threaded reading with `optimize_read_in_order = 1`. (It is my guess and good to check).

I think your guess is correct. I opened a related issue in #40675. When the data is sorted globally, `MergingSortedTransform` works like a concat processor and limit the reading to (almost) single thread.

> First idea
> Let's try to add buffering processor into query pipeline, after FilterTransform and before MergingSortedTransform. It will

> if LIMIT is not specified for MergingSortedTransform, allow reading more data and buffer it (e.g. up to max_block_size * 2 rows, or any other constant)
> if LIMIT is specified, the same, but formula will be like min(LIMIT, max_block_size * 2)

I implemented a prototype of this solution in #40675 and I think it works perfectly well. But there's a concern that if the result set is big, it can cause OOM.

**Comment by devcrafter** [comment]
@canhld94 https://github.com/ClickHouse/ClickHouse/issues/40675 is an issue, not a PR. What PR do you refer to?

**Comment by canhld94** [comment]
> @canhld94 https://github.com/ClickHouse/ClickHouse/issues/40675 is an issue, not a PR. What PR do you refer to?

I implemented it locally but didn't send a PR, just described how it was implemented in #40675. There's already a processor named QueueBuffer that we can put it before MergingSortedTransform, just need a little modification to support limit pushdown.

**Comment by UnamedRus** [comment]
Related https://github.com/ClickHouse/ClickHouse/issues/11482

Maybe related? https://github.com/ClickHouse/ClickHouse/issues/17364
